### PR TITLE
gk: add functionality to flush flow tables

### LINKED ISSUE #43: Extending the GK-GT protocol to support flushing policy decisions
When a network administrator changes a policy in response to an attack, she may need to flush all policy decisions cached at the Gatekeeper servers associated to a given destination in order to quickly re-establish the network. Thus, we need to extend the protocol of the GK-GT unit to support these flush requests.

### PR DESCRIPTION
**Pull request by mengxiang0811**
This patch closes #43

### CODE REVIEW

**Comment by mengxiang0811** [comment]
This rebase incorporated Cody's suggestions.

File: gk/main.c [hunk 1]
```diff
@@ -850,6 +850,101 @@ add_ggu_policy(struct ggu_policy *policy,
...
+
+			if (dst != NULL) {
+				matched &= ip4_same_subnet(
+					dst->addr.ip.v4.s_addr,
+					fe->flow.f.v4.dst,
+					ip4_dst_mask.s_addr);
+			}
+		}
+
+		if (proto == ETHER_TYPE_IPv6) {
```
Comment (AltraMayor) [inline gk/main.c]: Combine with previous `if` statement using an `else`. Add `likely()`.

File: gk/main.c [hunk 2]
```diff
@@ -850,6 +850,101 @@ add_ggu_policy(struct ggu_policy *policy,
...
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		bool matched = true;
+		struct flow_entry *fe =
+			&instance->ip_flow_entry_table[index];
+
+		if (proto != fe->flow.proto)
+			goto next;
+
+		if (proto == ETHER_TYPE_IPv4) {
```
Comment (AltraMayor) [inline gk/main.c]: Combine with previous `if` statement using an `else`.

File: gk/main.c [hunk 3]
```diff
@@ -850,6 +850,101 @@ add_ggu_policy(struct ggu_policy *policy,
...
+				instance->ip_flow_hash_table, fe);
+
+next:
+		index = rte_hash_iterate(instance->ip_flow_hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	GK_LOG(NOTICE,
+		"The GK block finished flushing flow table at %s with lcore %u\n",
+		__func__, rte_lcore_id());
```
Comment (AltraMayor) [inline gk/main.c]: This log entry would be more helpful if it included the number of flow entries removed.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: gk/main.c [hunk 4]
```diff
@@ -850,6 +850,94 @@ add_ggu_policy(struct ggu_policy *policy,
...
+	index = rte_hash_iterate(instance->ip_flow_hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		bool matched = true;
+		struct flow_entry *fe =
+			&instance->ip_flow_entry_table[index];
+
+		if (proto != fe->flow.proto)
+			goto next;
+		else if (proto == ETHER_TYPE_IPv4) {
```
Comment (AltraMayor) [inline gk/main.c]: I know that I've asked for it before, but drop the `else` and add a blank line after the `goto next;`. Sorry for the confusion.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

File: gk/main.c [hunk 5]
```diff
@@ -850,6 +850,94 @@ add_ggu_policy(struct ggu_policy *policy,
...
+	void *data;
+	struct in_addr ip4_src_mask;
+	struct in_addr ip4_dst_mask;
+	struct in6_addr ip6_src_mask;
+	struct in6_addr ip6_dst_mask;
+
+	memset(&ip4_src_mask, 0, sizeof(ip4_src_mask));
+	memset(&ip4_dst_mask, 0, sizeof(ip4_dst_mask));
+	memset(&ip6_src_mask, 0, sizeof(ip6_src_mask));
+	memset(&ip6_dst_mask, 0, sizeof(ip6_dst_mask));
```
[the remark this answers is not in the document]
Comment (mengxiang0811) [reply gk/main.c]: If we removed the ``memset()`` calls on the IPv6 masks, there will be an error: ``error: ‘*((void *)&ip6_dst_mask+8)’ may be used uninitialized in this function [-Werror=maybe-uninitialized]``

File: gk/main.c [hunk 6]
```diff
@@ -850,6 +850,91 @@ add_ggu_policy(struct ggu_policy *policy,
...
+	int32_t index;
+	uint32_t num_flushed_flows = 0;
+	const struct ip_flow *key;
+	void *data;
+	struct in_addr ip4_src_mask;
+	struct in_addr ip4_dst_mask;
+	struct in6_addr ip6_src_mask;
+	struct in6_addr ip6_dst_mask;
+
+	memset(&ip6_src_mask, 0, sizeof(ip6_src_mask));
```
Comment (AltraMayor) [inline gk/main.c]: Add `RTE_VERIFY(src->addr.proto == dst->addr.proto);` just before the `memset()`.

File: gk/main.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (AltraMayor) [reply gk/main.c]: Ok, to keep the code regular, let's have the `memset()`s under `if (src->addr.proto == ETHER_TYPE_IPv4) {`, and add `memset()`s for IPv4 masks under `} else if (likely(src->addr.proto == ETHER_TYPE_IPv6)) {`.

File: gk/main.c [hunk 7]
```diff
@@ -850,6 +850,91 @@ add_ggu_policy(struct ggu_policy *policy,
...
+
+		if (proto == ETHER_TYPE_IPv4) {
+			if (src->len != 0) {
+				matched &= ip4_same_subnet(
+					src->addr.ip.v4.s_addr,
+					fe->flow.f.v4.src,
+					ip4_src_mask.s_addr);
+			}
+
+			if (dst->len != 0) {
```
Comment (AltraMayor) [inline gk/main.c]: Change the test to `matched && dst->len != 0`.

This comment also applies to the IPv6 case.

File: gk/main.c [hunk 8]
```diff
@@ -850,6 +850,91 @@ add_ggu_policy(struct ggu_policy *policy,
...
+		bool matched = true;
+		struct flow_entry *fe =
+			&instance->ip_flow_entry_table[index];
+
+		if (proto != fe->flow.proto)
+			goto next;
+
+		if (proto == ETHER_TYPE_IPv4) {
+			if (src->len != 0) {
+				matched &= ip4_same_subnet(
```
Comment (AltraMayor) [inline gk/main.c]: Replace `&=` with `=`.

This comment also applies to the IPv6 case.

File: gk/main.c [hunk 9]
```diff
@@ -850,6 +850,91 @@ add_ggu_policy(struct ggu_policy *policy,
...
+					src->addr.ip.v4.s_addr,
+					fe->flow.f.v4.src,
+					ip4_src_mask.s_addr);
+			}
+
+			if (dst->len != 0) {
+				matched &= ip4_same_subnet(
```
Comment (AltraMayor) [inline gk/main.c]: Replace `&=` with `=`.

This comment also applies to the IPv6 case.

**Comment by AltraMayor** [comment]
Hi @mengxiang0811,

Could you squash my patches into yours and test the code? You can ignore my comments in the patches while squashing them. I wrote those comments for you to quickly understand what I changed.

**Comment by mengxiang0811** [comment]
This rebase is ready for another review.

**Pull request merged**